libr/anal/xrefs.c

```diff
@@ -195,333 +143,131 @@ R_API int r_anal_xrefs_set (RAnal *anal, const RAnalRefType type, ut64 from, ut6
-#endif
+	listxrefs (anal->dict_refs, from, list);
 	if (r_list_empty (list)) {
 		r_list_free (list);
 		list = NULL;
 	}
 	return list;
 }
 
-R_API RList *r_anal_xrefs_get_from (RAnal *anal, ut64 to) {
-	RList *list = r_list_newf (NULL);
+R_API RList *r_anal_xrefs_get_from(RAnal *anal, ut64 to) {
+	RList *list = r_anal_ref_list_new ();
 	if (!list) {
 		return NULL;
 	}
-#if USE_DICT
-	listxrefs (anal->dict_xrefs, to, list);
-	//listrefs (anal->dict_refs, to, list);
-#else
-	r_anal_xrefs_from (anal, list, "ref", R_ANAL_REF_TYPE_NULL, to);
-	r_anal_xrefs_from (anal, list, "ref", R_ANAL_REF_TYPE_CODE, to);
-	r_anal_xrefs_from (anal, list, "ref", R_ANAL_REF_TYPE_CALL, to);
-	r_anal_xrefs_from (anal, list, "ref", R_ANAL_REF_TYPE_DATA, to);
-	r_anal_xrefs_from (anal, list, "ref", R_ANAL_REF_TYPE_STRING, to);
-#endif
+	listxrefs (anal->dict_refs, to, list);
 	if (r_list_empty (list)) {
 		r_list_free (list);
 		list = NULL;
 	}
 	return list;
 }
 
-R_API bool r_anal_xrefs_init(RAnal *anal) {
-	sdb_reset (DB);
-	if (DB) {
-		sdb_array_set (DB, "types", -1, "code.jmp,code.call,data.mem,data.string", 0);
-		return true;
-	}
-	return false;
-}
-
-static int xrefs_list_cb_rad(RAnal *anal, const char *k, const char *v) {
-	ut64 dst, src = r_num_get (NULL, v);
-	if (!strncmp (k, "ref.", 4)) {
-		const char *p = r_str_rchr (k, NULL, '.');
-		if (p) {
-			dst = r_num_get (NULL, p + 1);
-			anal->cb_printf ("ax 0x%"PFMT64x" 0x%"PFMT64x"\n", src, dst);
-		}
+R_API void r_anal_xrefs_list(RAnal *anal, int rad) {
+	bool is_first = true;
+	RListIter *iter;
+	RAnalRef *ref;
+	RList *list = r_anal_ref_list_new();
+	listxrefs (anal->dict_xrefs, UT64_MAX, list);
+	if (rad == 'j') {
+		anal->cb_printf ("{");
 	}
-	return 1;
-}
-
-static int xrefs_list_cb_quiet(RAnal *anal, const char *k, const char *v) {
-	ut64 dst, src = r_num_get (NULL, v);
-	if (!strncmp (k, "ref.", 4)) {
-		const char *p = r_str_rchr (k, NULL, '.');
-		if (p) {
-			dst = r_num_get (NULL, p + 1);
-			char * type = strchr (k, '.');
-			if (type) {
-				type = strdup (type + 1);
-				char *t = strchr (type, '.');
-				if (t) {
-					*t = ' ';
-				}
-				char *T = (char *)r_str_rchr (type, NULL, '.');
-				if (T) {
-					T = (char *)r_str_rchr (T, NULL, '.');
-					if (T) {
-						*T = 0;
-						anal->cb_printf ("0x%08"PFMT64x" -> 0x%08"PFMT64x"  %s\n", src, dst, type);
-					}
+	r_list_foreach (list, iter, ref) {
+		int t = ref->type ? ref->type: ' ';
+		switch (rad) {
+		case '*':
+			anal->cb_printf ("ax%c 0x%"PFMT64x" 0x%"PFMT64x"\n",
+				t, ref->at, ref->addr);
+			break;
+		case '\0':
+			{
+				char *name = anal->coreb.getNameDelta (anal->coreb.core, ref->at);
+				r_str_replace_char (name, ' ', 0);
+				anal->cb_printf ("%40s", name? name: "");
+				free (name);
+				anal->cb_printf (" 0x%"PFMT64x" -> %9s -> 0x%"PFMT64x, ref->at, analref_toString (t), ref->addr);
+				name = anal->coreb.getNameDelta (anal->coreb.core, ref->addr);
+				r_str_replace_char (name, ' ', 0);
```

Comment:
null deref, you are checking for name being null after using it in r_str_replace_char()

Reply:
it was already there, I didn't touch nor looked at this code at all. I just assumed it was working, but i can change that in this PR too.

Reply:
this is actually fine too, because r_str_replace_char checks for NULL.

Reply:
ok